zkevm-circuits/src/evm_circuit/execution/return.rs

```diff
@@ -24,25 +46,313 @@ impl<F: Field> ExecutionGadget<F> for ReturnGadget<F> {
         let opcode = cb.query_cell();
         cb.opcode_lookup(opcode.expr(), 1.expr());
 
-        // TODO: Other constraints are ignored now for RETURN to serve as a
-        // mocking terminator
+        let offset = cb.query_cell();
+        let length = cb.query_rlc();
+        cb.stack_pop(offset.expr());
+        cb.stack_pop(length.expr());
+        let range = MemoryAddressGadget::construct(cb, offset, length);
+
+        let is_success = cb.call_context(None, CallContextFieldTag::IsSuccess);
+        cb.require_equal(
+            "if is_success, opcode is RETURN. if not, opcode is REVERT",
+            opcode.expr(),
+            is_success.expr() * OpcodeId::RETURN.expr()
+                + not::expr(is_success.expr()) * OpcodeId::REVERT.expr(),
+        );
+
+        // There are 4 cases non-mutually exclusive, A to D, to handle, depending on if
+        // the call is, or is not, a create, root, or successful. See the specs at
+        // https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/opcode/F3RETURN.md
+        // for more details.
+        let is_create = cb.curr.state.is_create.expr();
+        let is_root = cb.curr.state.is_root.expr();
+
+        // These are globally defined because they are used across multiple cases.
+        let copy_rw_increase = cb.query_cell();
+        let copy_rw_increase_is_zero = IsZeroGadget::construct(cb, copy_rw_increase.expr());
+
+        // Case A in the specs.
+        cb.condition(is_create.clone() * is_success.expr(), |cb| {
+            cb.require_equal(
+                "increase rw counter once for each memory to bytecode byte copied",
+                copy_rw_increase.expr(),
+                range.length(),
+            );
+        });
+        cb.condition(
+            is_create.clone() * is_success.expr() * not::expr(copy_rw_increase_is_zero.expr()),
+            |_cb| {
+                // TODO: copy_table_lookup for contract creation.
+            },
+        );
 
-        Self { opcode }
+        // Case B in the specs.
+        cb.condition(is_root.expr(), |cb| {
+            cb.require_next_state(ExecutionState::EndTx);
+            cb.call_context_lookup(
+                false.expr(),
+                None,
+                CallContextFieldTag::IsPersistent,
+                is_success.expr(),
+            );
+        });
+
+        // Case C in the specs.
+        // TODO: have copy_table_lookup update rw_counter expression so that this can go
+        // at the end of the constraints.
+        let restore_context = cb.condition(not::expr(is_root.expr()), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                is_success.expr(),
+                2.expr() + copy_rw_increase.expr(),
```

Comment:
I wonder where this `2` comes from.  I'm counting 3 + (2) fixed lookups:
- 2 Stack lookup
- CallContextFieldTag::Success lookup
- CallContextFieldTag::{ReturnDataOffset, ReturnDataLength} if not(is_create)

Could you help me clarify this?  Considering that the tests are passing, maybe I missed something :thinking:

Reply:
I renamed this argument to be `subsequent_rw_lookups` because it's the expression for the increase in the rw counter after this gadget is constructed. The two come from the lookups for `ReturnDataOffset` and `ReturnDataLength` in case D (if it's not a create, which was not accounted for).

Reply:
Ah got it, thanks!